compiler/noirc_evaluator/src/ssa/checks/check_for_underconstrained_values.rs

```diff
@@ -63,6 +92,335 @@ fn check_for_underconstrained_values_within_function(
+                            }
+                        },
+                        Value::Function(callee) => match all_functions[&callee].runtime() {
+                            RuntimeType::Brillig(_) => {
+                                // Record arguments/results for each Brillig call for the check
+                                trace!("Brillig function {} called at {}", callee, instruction);
+                                self.tainted.insert(
+                                    *instruction,
+                                    BrilligTaintedIds::new(&arguments, &results),
+                                );
+                            }
+                            RuntimeType::Acir(..) => {
+                                // Record all the function arguments as parents of the results
+                                self.update_children(&arguments, &results);
+                            }
+                        },
+                        Value::ForeignFunction(..) => {
+                            debug!("should not be able to reach foreign function from non-Brillig functions, {func_id} in function {}", function.name());
+                        }
+                        Value::Instruction { .. }
+                        | Value::NumericConstant { .. }
+                        | Value::Param { .. } => {
+                            debug!(
+                                "should not be able to call {func_id} in function {}",
+                                function.name()
+                            );
+                        }
+                    }
+                }
+                // For array get operations, we link the resulting values to
+                // the corresponding array value ids
+                // (this is required later because for now we consider array elements
+                // being constrained as valid as the whole arrays being constrained)
+                Instruction::ArrayGet { array, .. } => {
+                    for result in &results {
+                        self.array_elements.insert(*result, function.dfg.resolve(*array));
+                    }
+                    // Record all the used arguments as parents of the results
+                    self.update_children(&arguments, &results);
+                }
+                Instruction::ArraySet { .. }
+                | Instruction::Binary(..)
+                | Instruction::Cast(..)
+                | Instruction::IfElse { .. }
+                | Instruction::Not(..)
+                | Instruction::Truncate { .. } => {
+                    // Record all the used arguments as parents of the results
+                    self.update_children(&arguments, &results);
+                }
+                // These instructions won't affect the dependency graph
+                Instruction::Allocate { .. }
+                | Instruction::DecrementRc { .. }
+                | Instruction::EnableSideEffectsIf { .. }
```

Comment:
Could using a value as the condition in `EnableSideEffectsIf` be considered a constraint?

Reply:
@Rumata888 what do you think?

Reply:
No idea. Could you give an example of a specific case where it is used?

Reply:
`EnableSideEffectsIf` is a marker instruction to mark where `if` blocks were once they were flattened and removed. In pseudo code
```noir
if foo {
    print("foo");
} else {
    print("!foo");
}
```
would translate to the pseudo-ssa post-flattening:
```
enable_side_effects_if foo
call print("foo")
v0 = not foo
enable_side_effects_if v0
call print("!foo")
enable_side_effects_if true
... rest of program ...
```
It is mostly used to track whether we should actually perform or ignore calls since in our target there are no blocks or control-flow. Other instructions like `constrain` may have their own mechanisms for this. In the case of constrain we can multiply the constrain value by the current condition, e.g. `constrain a == b` becomes `constrain a*cond == b*cond` which will zero both out if `cond` is false because we're in a branch that wasn't taken.

Reply:
Doesn't seem like a valid constraint in terms of this check then. @Rumata888 do you agree?

Reply:
Unfortunately, I think we should treat it as one and mark all the results of instructions affected by a condition that uses brillig output as also tainted by brillig output. I'd argue that it doesn't work the same for the arguments of the brillig function, since it doesn't constrain them. However, brillig function calls could be used to enable a particular constraint.

Reply:
Alright. This should go into a separate issue then.